src/expression.js

```diff
@@ -705,17 +705,24 @@ pp.parseFunctionBody = function(node, isArrowFunction) {
     // Start a new scope with regard to labels and the `inFunction`
     // flag (restore them to their old value afterwards).
     let oldLabels = this.labels
+    let oldVarDeclaredNames = this.varDeclaredNames
+    this.varDeclaredNames = new Set()
     this.labels = []
     if (useStrict) this.strict = true
-    node.body = this.parseBlock(true)
+
+    // Add the params to varDeclaredNames to ensure that an error is thrown
+    // if a let/const declaration in the function clashes with one of the params.
+    node.params.forEach(param => this.checkLVal(param, "var"))
```

Comment:
Callers will have either parsed the params with `parseBindingList`, or already checked them with `toAssignable`, so checking them a second time seems wasteful.